### ISSUE TITLE
Dataset, Sampler, and Dataloader in distributed training/infernce using dp/ddp strategies 

### ISSUE THREAD

**Issue by mrezasoltani** [lhotse-speech/lhotse#976]
Hi,

I am running to many issues when I want to use K2SpeechRecognitionDataset with different sampler methods in lhotse and distributed training/inference strategies such as dp/ddp. To be more specific, I have created a cutset for my own dataset, including 60000 cuts (trimmed to supervisions) with different durations. 

Firstly, I tried to do inference on a pre-trained medium-size whisper (from OpenAI) model. This works fine if I use only single GPU. Essentially, I do the followings:

dataset = K2SpeechRecognitionDataset(return_cuts=True, input_strategy=AudioSamples())
batch_size = 8
sampler = DynamicBucketingSampler(cuts, shuffle=False, max_cuts=batch_size,
                                               max_duration=float('inf'), num_buckets=3)
dloader = torch.utils.data.DataLoader(
    dataset,
    sampler=sampler,
    batch_size=None,
    num_workers=16)
 
### checkpoint is the location of downloaded pre-trained whisper model ###
model.load_state_dict(checkpoint)  
model.eval()
model.to("cuda")

### whisper needs feature extraction first (e.g., mel features) ####
def forward(batch, model):
      arr = whisper.pad_or_trim(batch["inputs"])
      mel = whisper.log_mel_spectrogram(arr).to("cuda")
      return  whisper.decode(self.model, mel, whisper.DecodingOptions())
    
res = []
for batch in dloader:
    res.append(forward(batch, model))

Now, I need to scale things up. That is, I want to use 2 or more GPUs in one machine or across multiple machines. To just start with 2 GPUs in one machine and using Pytorch-lightning, I use exactly the above dataset and sampler method and dp (DataParallel) strategy. However, I get the following error. 

"ValueError: A frozen dataclass was passed to `apply_to_collection` but this is not allowed."

To give you a littile more context, this is the last lines of the error:

> File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1190, in _call_impl
>     return forward_call(*input, **kwargs)
>   File "/usr/local/lib/python3.8/dist-packages/pytorch_lightning/overrides/data_parallel.py", line 85, in forward
>     output = apply_to_collection(output, dtype=(numbers.Number, Tensor), function=output_transform)
>   File "/usr/local/lib/python3.8/dist-packages/lightning_utilities/core/apply_func.py", line 69, in apply_to_collection
>     v = apply_to_collection(
>   File "/usr/local/lib/python3.8/dist-packages/lightning_utilities/core/apply_func.py", line 105, in apply_to_collection
>     raise ValueError(
> ValueError: A frozen dataclass was passed to `apply_to_collection` but this is not allowed.

Even with one GPU, using the same dp strategy, and SimpleCutSampler (please see below)), I get the above error (I cannot use SpeechRecognitionDataset(cuts) as my cuts do not have any features.):

dataset = SpeechRecognitionDataset(return_cuts=True, input_strategy=AudioSamples())
sampler = SimpleCutSampler(cutsegs, shuffle=False, max_cuts=batch_size, max_duration=float('inf'))

Also, I changed the sampler method in lhotse as follows:

sampler = DynamicBucketingSampler(
            cutsegs, shuffle=False, max_duration=80.0, num_buckets=10)

With the above sampler method (1 or 2 GPUS with the same dp strategy), I get the "frozen dataclass error" again. 

Doing more tests, with ddp strategy (recommended by Pytorch and Pytorch-lightning even for a single machine setup), I get the following error (using 1 or 2 GPUs with DynamicBucketingSampler which is initialized by a finite max_duration, e.g., 80.0 s):

> TypeError: You seem to have configured a sampler in your DataLoader which does not provide `__len__` method. The sampler was about to be replaced by `DistributedSamplerWrapper` since `replace_sampler_ddp` is True and you are using distributed training. Either provide `__len__` method in your sampler, remove it from DataLoader or set `replace_sampler_ddp=False` if you want to handle distributed sampling yourself.

If I set "replace_sampler_ddp=False", I get the "A frozen dataclass" error. Moreover, I cannot remove sampler from dataloader since I get the following error:
"TypeError: object of type 'K2SpeechRecognitionDataset' has no len()"

To sum up, I am not sure if above errors are because of conflict between Pytorch-lightning and lhotse, or something else is going on. I have to say that all the above tests have been done for only inference and no training/fine-tuning has been done. 

Can you please provide me some hints?

Thanks.

**Comment by pzelasko** [comment]
I don’t know why you are experiencing the “frozen dataclass” issue, it comes from PyTorch Lightning and seems more related to that library. 

You can check out Icefall recipes for an example of a training script that works seamlessly with Lhotse and DDP, and adapt that approach to your use case. Eg: https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless2/train.py

**Comment by mrezasoltani** [comment]
Thanks for your answer. Can you have a look at the following error (I don't use any more Pytorch-Lightning):

> File "/usr/local/lib/python3.8/dist-packages/lhotse/dataset/sampling/dynamic_bucketing.py", line 146, in __init__
>     self.duration_bins = estimate_duration_buckets(
>   File "/usr/local/lib/python3.8/dist-packages/lhotse/dataset/sampling/dynamic_bucketing.py", line 277, in estimate_duration_buckets
>     durs = np.array([c.duration for c in cuts])
>   File "/usr/local/lib/python3.8/dist-packages/lhotse/dataset/sampling/dynamic_bucketing.py", line 277, in <listcomp>
>     durs = np.array([c.duration for c in cuts])
>   File "/usr/local/lib/python3.8/dist-packages/lhotse/lazy.py", line 165, in values
>     yield from self
>   File "/usr/local/lib/python3.8/dist-packages/lhotse/lazy.py", line 443, in __iter__
>     for cuts in self.iterator:
>   File "/usr/local/lib/python3.8/dist-packages/lhotse/lazy.py", line 165, in values
>     yield from self
>   File "/usr/local/lib/python3.8/dist-packages/lhotse/lazy.py", line 216, in __iter__
>     yield from map(deserialize_item, self.source)
>   File "/usr/local/lib/python3.8/dist-packages/lhotse/lazy.py", line 186, in __iter__
>     for line in f:
>   File "/usr/lib/python3.8/gzip.py", line 305, in read1
>     return self._buffer.read1(size)
>   File "/usr/lib/python3.8/_compression.py", line 68, in readinto
>     data = self.read(len(byte_view))
>   File "/usr/lib/python3.8/gzip.py", line 479, in read
>     if not self._read_gzip_header():
>   File "/usr/lib/python3.8/gzip.py", line 427, in _read_gzip_header
>     raise BadGzipFile('Not a gzipped file (%r)' % magic)
> gzip.BadGzipFile: Not a gzipped file (b'\xf3\x00')

**I am using DDP strategy; however, it complains of the sampler which is set to** 
sampler = DynamicBucketingSampler(cutsegs, shuffle=False, max_duration=10.0, num_buckets=10)

**Comment by pzelasko** [comment]
That looks like your cutset file could be corrupted, is it possible the script that created it crashed? Are you able to unzip it in the CLI and/or load it normally otherwise? You can check with e.g. `gunzip my-cut-set.jsonl.gz`. If that's the case, you'll need to re-create it.

**Comment by mrezasoltani** [comment]
> That looks like your cutset file could be corrupted, is it possible the script that created it crashed? Are you able to unzip it in the CLI and/or load it normally otherwise? You can check with e.g. `gunzip my-cut-set.jsonl.gz`. If that's the case, you'll need to re-create it.

Thanks for your help. I had set the random_ids flag to True in CutSet.from_manifests. When I set it to False, I was able to run it.
